src/wallet/sqlite.cpp

```diff
@@ -265,22 +265,106 @@ void SQLiteBatch::Close()
 {
-    return false;
+    if (!m_database.m_db) return false;
+    assert(m_read_stmt);
+
+    // Bind: leftmost parameter in statement is index 1
+    int res = sqlite3_bind_blob(m_read_stmt, 1, key.data(), key.size(), SQLITE_STATIC);
+    if (res != SQLITE_OK) {
+        sqlite3_clear_bindings(m_read_stmt);
+        sqlite3_reset(m_read_stmt);
+        return false;
+    }
+    res = sqlite3_step(m_read_stmt);
+    if (res != SQLITE_ROW) {
+        sqlite3_clear_bindings(m_read_stmt);
+        sqlite3_reset(m_read_stmt);
+        return false;
+    }
+    // Leftmost column in result is index 0
+    const char* data = (const char*)sqlite3_column_blob(m_read_stmt, 0);
+    int data_size = sqlite3_column_bytes(m_read_stmt, 0);
+    value.write(data, data_size);
+
+    sqlite3_clear_bindings(m_read_stmt);
+    sqlite3_reset(m_read_stmt);
+    return true;
 }
 
 bool SQLiteBatch::WriteKey(CDataStream&& key, CDataStream&& value, bool overwrite)
 {
-    return false;
+    if (!m_database.m_db) return false;
+    if (m_read_only) throw std::runtime_error("Write called on database in read-only mode");
+    assert(m_insert_stmt && m_overwrite_stmt);
+
+    sqlite3_stmt* stmt;
+    if (overwrite) {
+        stmt = m_overwrite_stmt;
+    } else {
+        stmt = m_insert_stmt;
+    }
+
+    // Bind: leftmost parameter in statement is index 1
+    // Insert index 1 is key, 2 is value
+    int res = sqlite3_bind_blob(stmt, 1, key.data(), key.size(), SQLITE_STATIC);
+    if (res != SQLITE_OK) {
+        sqlite3_clear_bindings(stmt);
+        sqlite3_reset(stmt);
+        return false;
+    }
+    res = sqlite3_bind_blob(stmt, 2, value.data(), value.size(), SQLITE_STATIC);
+    if (res != SQLITE_OK) {
+        sqlite3_clear_bindings(stmt);
+        sqlite3_reset(stmt);
+        return false;
+    }
+
+    // Execute
+    res = sqlite3_step(stmt);
+    sqlite3_clear_bindings(stmt);
+    sqlite3_reset(stmt);
+    return res == SQLITE_DONE;
 }
 
 bool SQLiteBatch::EraseKey(CDataStream&& key)
 {
-    return false;
+    if (!m_database.m_db) return false;
+    if (m_read_only) throw std::runtime_error("Erase called on database in read-only mode");
+    assert(m_delete_stmt);
+
+    // Bind: leftmost parameter in statement is index 1
+    int res = sqlite3_bind_blob(m_delete_stmt, 1, key.data(), key.size(), SQLITE_STATIC);
+    if (res != SQLITE_OK) {
+        sqlite3_clear_bindings(m_delete_stmt);
+        sqlite3_reset(m_delete_stmt);
+        return false;
+    }
+
+    // Execute
+    res = sqlite3_step(m_delete_stmt);
+    sqlite3_clear_bindings(m_delete_stmt);
+    sqlite3_reset(m_delete_stmt);
+    return res == SQLITE_DONE;
 }
 
 bool SQLiteBatch::HasKey(CDataStream&& key)
 {
-    return false;
+    if (!m_database.m_db) return false;
+    assert(m_read_stmt);
+
+    // Bind: leftmost parameter in statement is index 1
+    bool ret = false;
+    int res = sqlite3_bind_blob(m_read_stmt, 1, key.data(), key.size(), SQLITE_STATIC);
```

Comment:
40b15923da624272830ab51ac30be892b9d5712b

It was already suggested in https://github.com/bitcoin/bitcoin/pull/19077#pullrequestreview-497108156

> Could add a dedicated statement for HasKey, see https://stackoverflow.com/a/9756276/1978589. Are you going to add any index to main table?

Anyway it could also log "Unable to bind value to statement" when binding fails.